Fixed Issue 35 (use linked lists in JS)

frank-emrich opened the pull request:
One thing I'm not terribly happy about is the fact that I currently represent Nil as [] (i.e., the empty Javascript Array). This is mostly due to the fact that it's short and easy to parse. If we used something different, like a dedicated string or the empty object {}, it would be kind of hacky to parse this to the empty list in jsonparse.mly.

In addition, I could have squashed a few commits. In particular, what I did in 
bed80f3 was later mostly reverted. Do we pay a lot of attentition to squashing such useless commits?

jamescheney commented:
We normally do "squash and merge" or, if necessary, "rebase and merge" (or manually rebasing for those with sufficient git-fu, unlike me).   See:

https://github.com/links-lang/links/wiki/Working-conventions

dhil commented:
> One thing I'm not terribly happy about is the fact that I currently represent Nil as [] (i.e., the empty Javascript Array).

Why not use `null`? Using `[]` to denote the empty list is somewhat troublesome as `[] === []` and `[] == []` are both `false`, whereas `null === null` and `null == null` are both `true` in JavaScript.

Also using `[]` as nil means that (source level) homogeneous lists would be represented by heterogeneous lists at runtime which may lead to all kinds of strangeness if we are not careful.

frank-emrich commented:
@jamescheney I tested all the examples now and fixed the additional bugs I found. I also tested with --enable-handlers as suggested by Daniel. I found out that one example wasnt't working in the current master when --enable-handers is set, and created a separate issue for that.

@dhil null was previously used for Nil. I thought that may not be ideal since we then have to parse null to the empty list in jsonparse.mly, but then I realized that currently null is parsed to the empty record & there's a comment saying that raising an error may be better at this point. So I guess just sticking with null is okay, I changed it back to null now.

jamescheney commented:
This looks reasonable to me.  Maybe @slindley should look at it when he gets back too?  I seem to recall that the original issue claimed that using linked lists would avoid some bad performance associated with arrays, e.g. sorting becoming quadratic.  Has this been checked?

frank-emrich commented:
In general, calling tail() on a list involved copying the whole remainder in the old JS code. This is not the case any longer. But I don't know of any particular example I could use to see if there is a measurable performance benefit resulting from my changes.

Some of the runtime machinery in jslib.js still uses JS Arrays, but only for internal purposes. I am cautiously optimistic that I eliminated all occurences where such runtime stuff could "leak" into Links programs. This involved adapting some of the code that deals with XML representations of DOM nodes, but that should work now.

jamescheney commented:
It looks like this is ready to merge.  I'll do that Monday if no one objects.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
-  return out;
-}
-const lsAt = _lsAt;
+    take: function (n, xs) {
```
2) With the new (uniform) list representation I think this would be defined in the prelude rather as a built-in function. To the best of my knowledge the Links compiler does not depend on `take`.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
-	return xs === null;
-}
-const lsEmpty = _lsEmpty;
+    drop: function(n, xs) {
```
Replace `take` with `drop` in 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
-  return _lsFromArray(arr);
-}
-const lsZip = _lsZip;
+    length: function (xs) {
```
As with 2) I think this can be user-defined now.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
-  return currentMin;
-}
-const lsMinimum = _lsMinimum;
+    last: function(xs) {
```
I think this can be user-defined as well.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+      return out;
+    },
+
+    empty: function(xs) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+      return rootEl;
+    },
+
+    at: function(xs, i) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+      return LINKEDLIST.isNil(xs);
+    },
+
+    zip: function(xs, ys) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+      return lst;
+    },
+
+    replicate: function(n, item) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+      return LINKEDLIST.lsFromArray(arr);
+    },
+
+    range:  function(a, b) {
```
Same like a specialised version of `intRange` which we have special sugar for in the source syntax, i.e. `[1...42]`. I wonder whether `range` is used anywhere?

frank-emrich replied:
I haven't found any occurences

dhil replied:
Interesting, maybe it is in fact dead code.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+    },
+
+    // faster than _replicate? not really?
+    repeat: function(n, x) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+  // in order to be able to handle this kind of situation
+  // more robustly)
+
+    min: function(xs) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+      return currentMin;
+    },
+
+    max: function(xs) {
```
Same as 2)

dhil commented on lib/js/jslib.js:
```diff
@@ -1269,6 +1359,12 @@ const LINKS = new (function() {
+        TYPES.isLinkedList(r) && r != null) {
+      return LINKEDLIST.equal(l,r)
+
+    return true;
```
Dead code, this return statement can never be reached.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
+    //Use this to check if something is Nil (using == or === may fail, since we have e.g. that  [] === [] is false in JavaScript)
+    isNil: function(n) {return n === null},
+  
+    Cons: function(x,xs) {DEBUG.assert(LINKEDLIST.isLinkedList(xs), "Second argument must be linked list"); return {_head: x, _tail: xs}},
```
**Nitpick** why is `Cons` capitalised?

frank-emrich replied:
same as for Nil

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
 
-var lsNil = null;
+    //Using [] here for easier parsing (if we used e.g. a string, JSON parsing would become ambiguous)
+    Nil: null,
```
**Nitpick** why is `Nil` capitalised?

frank-emrich replied:
It was like that before, I guess the original author wanted to mimick the "capitalize data constructors" rule found in languages like Haskell?

slindley replied:
I'm not sure if this is related, but some functions were capitalised at some point in order to ensure that there could never be a clash with user defined functions (which must always begin with lower case letters). There is a long outstanding bug with pattern matching compilation whereby one can witness strange behaviour by redefining hd and tl (collections.tests: Case patterns (with redefined hd)). It would make sense to address these name hygiene problems more systematically.

dhil commented on lib/js/jslib.js:
```diff
@@ -1,187 +1,255 @@
 'use strict';
 
 /* LIST MANIPULATING FUNCTIONS */
+const LINKEDLIST =  new function(){
```
Please avoid using `new function` as it suffers from potential security and performance issues. It is somewhat similar to be using `eval` to dynamically create and execute arbitrary code. Rather use 
```
const LINKEDLIST = (function() {
   return ....;
}());
```

jamescheney commented:
I'll defer merging until the requested changes are addressed.

frank-emrich commented:
I have pushed the requested changes

jamescheney commented:
Is this now ready to go?  Let's discuss tomorrow.

frank-emrich commented:
The current status now fixes all of the issues raised by Daniel except the removal of the hard-coded list functions. I've created #298 for this.
This PR should now be ready to merge.

jamescheney commented:
@dhil, is there anything left in your code review that should be fixed at this point (i.e. not deferred to #298)

jamescheney commented:
Manually dismissing @dhil's review, will merge at the end of today.

jamescheney commented:
Merged.  Whew!  Thanks @frank-emrich !
